File: backend/schemas/repository.py
from datetime import datetime

from pydantic import BaseModel, Field, HttpUrl


class RepositoryCreate(BaseModel):
    url: HttpUrl = Field(description="Public GitHub repository URL")


class RepositoryResponse(BaseModel):
    id: int
    url: str
    full_name: str
    provider: str
    status: str
    clone_path: str
    default_branch: str | None
    created_at: datetime
    updated_at: datetime


class RepositoryListResponse(BaseModel):
    repositories: list[RepositoryResponse]


class SymbolResponse(BaseModel):
    id: int
    name: str
    kind: str
    start_line: int
    end_line: int
    signature: str | None
    parent_name: str | None
    file_path: str
    language: str


class SkippedFileResponse(BaseModel):
    path: str
    reason: str


class RepositoryParseSummaryResponse(BaseModel):
    repository_id: int
    file_count: int
    symbol_count: int
    skipped_count: int
    by_language: dict[str, int]
    by_kind: dict[str, int]
    by_skip_reason: dict[str, int]
    skipped_files: list[SkippedFileResponse]
    skipped_returned: int
    skipped_limit: int


class RepositorySymbolsResponse(RepositoryParseSummaryResponse):
    symbols: list[SymbolResponse]
    symbols_returned: int
    symbols_limit: int


class EmbeddingSummaryResponse(BaseModel):
    repository_id: int
    vector_count: int
    bm25_chunk_count: int = 0
    hybrid_ready: bool = False
    graph_node_count: int = 0
    graph_edge_count: int = 0
    graph_ready: bool = False

File: frontend/src/types/repository.ts
export type Repository = {
  id: number
  url: string
  full_name: string
  provider: string
  status: string
  clone_path: string
  default_branch: string | null
  created_at: string
  updated_at: string
}

export type SkippedFile = {
  path: string
  reason: string
}

export type RepositoryParseSummary = {
  repository_id: number
  file_count: number
  symbol_count: number
  skipped_count: number
  by_language: Record<string, number>
  by_kind: Record<string, number>
  by_skip_reason: Record<string, number>
  skipped_files: SkippedFile[]
  skipped_returned: number
  skipped_limit: number
}

export type EmbeddingSummary = {
  repository_id: number
  vector_count: number
  bm25_chunk_count: number
  hybrid_ready: boolean
  graph_node_count: number
  graph_edge_count: number
  graph_ready: boolean
}

export type GraphSummary = {
  repository_id: number
  node_count: number
  edge_count: number
  nodes_by_type: Record<string, number>
  edges_by_type: Record<string, number>
  graph_ready: boolean
}

export type GraphDependency = {
  source_file: string
  target_label: string
  target_type: string
  edge_type: string
}

export type GraphDependencies = {
  repository_id: number
  dependency_count: number
  dependencies: GraphDependency[]
  limit: number
}
